Boj14003.kt

```diff
@@ -0,0 +1,68 @@
+package ps.ps
+
+/**
+ * @author : Unagi_zoso
+ * @date : 2023-10-20
+ */
+// https://www.acmicpc.net/problem/14003
+
+fun main() = System.`in`.bufferedReader().use { reader ->
+    System.out.bufferedWriter().use { writer ->
+        val n = reader.readLine().toInt()
+        val inputList = mutableListOf(0) + reader.readLine().split(' ').map { it.toInt() }
+        val indexList = MutableList(1000005) { 0 } // 각 inputList 원소들이 valueList의 몇 번 인덱스에 할당되었는지 저장된다.
+        val valueList = MutableList(1) { 0 }
+
+        makeLISListsWithInputList(indexList, valueList, inputList, n) // inputList를 통해 indexList, valueList를 만든다.
+
+        val lis = findLIS(indexList, inputList, valueList.size - 1).joinToString(" ") { it.toString() }
+        writer.write(
+            """
+            ${valueList.size - 1}
+            $lis
+            """.trimMargin(),
+        )
+    }
+}
+
+fun makeLISListsWithInputList(
+    indexList: MutableList<Int>,
+    valueList: MutableList<Int>,
+    inputList: List<Int>,
+    n: Int,
+) {
+    indexList[1] = 1
+    valueList.add(inputList[1])
+    for (inputIdx in 2..n) {
+        if (valueList[valueList.size - 1] < inputList[inputIdx]) { // 현재 input값이 valueList의 값보다 클 경우 value에 새 원소로 추가
+            valueList.add(inputList[inputIdx])
+            indexList[inputIdx] = valueList.size - 1 // indexList에는 input값이 저장된 valueList의 인덱스가 저장된다
+        } else { // valueList에서 현재 input값이 들어갈 수 있는 위치(인덱스)를 이분탐색으로 구한다.
+            val result = valueList.binarySearch(1, valueList.lastIndex, inputList[inputIdx])
+            valueList[result] = inputList[inputIdx]
+            indexList[inputIdx] = result
+        }
+    }
+}
+
+fun List<Int>.binarySearch(_left: Int, _right: Int, target: Int): Int {
```

Comment:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/binary-search.html
코틀린에서는 기본적으로 바이너리 서치를 구현해 두었는데
재 정의하신 이유가 있을까요??

Reply:
바이너리 서치 구현이 잘 기억이 안나 학습과 확장 함수 응용의 의미로 한 번 만들어보았네요.
앞으로는 환경에서 지원해주는 API를 잘 알아보고 사용하겠습니다!
감사합니다!

Reply:
코틀린 API 바이너리 서치를 좀 알아봤는데 다음과 같이 찾고자 하는 element가 없을 시 -insertion point - 1을 반환하네요.

 * @return the index of the found element, if it is contained in the list within the specified range;
 * otherwise, the inverted insertion point `(-insertion point - 1)`.

제가 구현한 바이너리 서치와 insertion point를 정하는데 차이가 있네요. 
코틀린의 바이너리 서치는 이 insertion point 정하는 방식을 외부에서 주입 받지 못해 변경도 불가능해서, 제가 구현한 바이너리 서치가 제 해결 방식에 조금 더 적합해 보입니다.

Reply:
기존에 제공되는 함수를 직접 구현해야하는 경우엔 그 이유를 주석이나 문서 등으로 남기는 편이 좋아보이네요.
감사합니다!

Reply:
```kotlin
            var result = valueList.binarySearch(1, valueList.lastIndex) {
                if (it < inputList[inputIdx]) -1 else 1
            }
            if (result < 0) {
                result = -result - 1
            }
```

요렇게 하면 기본제공되는 메소드로 같은 결과를 이끌어 낼수 있습니다~!

<img width="1161" alt="스크린샷 2023-10-22 오후 8 04 18" src="https://github.com/BI-kotlin-in-action/Juyeol-Song/assets/44064530/e273eefa-2276-4f94-96a9-8ca20e217f21">

Reply:
요것도 반영 부탁드릴게요~!

Reply:
이런 방법이 있었네요! 많이 배우는 거 같아요! 감사합니다!